gossip/src/main.rs

```diff
@@ -342,17 +320,40 @@ fn process_rpc_url(
     Ok(())
 }
 
+fn get_gossip_address(
+    matches: &ArgMatches,
+    entrypoint_addr: Option<SocketAddr>,
+) -> Result<SocketAddr, Box<dyn error::Error>> {
+    let gossip_host = parse_gossip_host(matches, entrypoint_addr);
+    let gossip_addr = SocketAddr::new(
+        gossip_host,
+        value_t!(matches, "gossip_port", u16).unwrap_or_else(|_| {
+            solana_net_utils::find_available_port_in_range(
+                IpAddr::V4(Ipv4Addr::UNSPECIFIED),
+                (0, 1),
+            )
+            .expect("unable to find an available gossip port")
+        }),
+    );
+    Ok(gossip_addr)
+}
+
 fn main() -> Result<(), Box<dyn error::Error>> {
     solana_logger::setup_with_default_filter();
 
     let matches = parse_matches();
     let socket_addr_space = SocketAddrSpace::new(matches.is_present("allow_private_addr"));
     match matches.subcommand() {
-        ("spy", Some(matches)) => {
-            process_spy(matches, socket_addr_space)?;
-        }
-        ("rpc-url", Some(matches)) => {
-            process_rpc_url(matches, socket_addr_space)?;
+        (gossip_type, Some(matches)) => {
+            let entrypoint_addr = parse_entrypoint(matches);
+            let gossip_addr = get_gossip_address(matches, entrypoint_addr)?;
+            match gossip_type {
+                "spy" => process_spy(matches, socket_addr_space, entrypoint_addr, gossip_addr)?,
+                "rpc-url" => {
+                    process_rpc_url(matches, socket_addr_space, entrypoint_addr, gossip_addr)?
+                }
+                _ => unreachable!(),
+            }
```

Comment:
As in the other comment, my inclination is not to add the logic here in `main` but just call the newly added helper function instead of the duplicate code.
That would avoid tying `spy` and `rpc-url` processing like here.